### ISSUE TITLE
Terrain altered by weather

### ISSUE THREAD

**Issue by mugling** [CleverRaven/Cataclysm-DDA#18274]
I'd like to start work on a system for weather affecting terrain. For example with sufficient rain grass should to muddy fields and standing water should become icy in the winter.

The question is how to write a reasonably performant implementation that dosen't become too blocky because of the submap design. Possibly a ring buffer that stores the last few months of weather and then applies them to adjacent submaps on load?

Suggestions?

**Comment by Coolthulhu** [comment]
Performance isn't a huge concern here:
* Solar panels generate a sum of weather conditions since last update with 1 hour granularity
* Ditto for funnels (both vehicle and stationary)
* Each food item gets a separate weather report for rot calculations, also 1 hour granularity

If per-(food)-item on-load calculations are not noticeable, per-submap ones shouldn't be either.

There are worse problems, though:
* Weather isn't very well balanced, so the simulation should be run not from turn 0, but extended deep into negatives, until it converges or reaches some cap
* Per-tile vs. per submap problem. The submaps should change gradually, but in a way that makes sense and doesn't expose submap bounds
* Retaining weather accumulation data that doesn't break with big time shifts. For example, if constant rain for 2 months produces 30 wet tiles, loading the submap after a month and then again after another shouldn't produce 15 or 60 wet tiles.

**Comment by mugling** [comment]
> Weather isn't very well balanced, so the simulation should be run not from turn 0, but extended deep into negatives, until it converges or reaches some cap

Can you rephrase?

> but in a way that makes sense and doesn't expose submap bounds

This is my main concern

**Comment by Coolthulhu** [comment]
> Can you rephrase?

If you just let x amount of rain turn terrain into mud and y amount of "dryness" turn it into desert, there's a high chance that the entire map will be mud by second season.
By running the simulation for a set amount of time backwards, the first day will have roughly the same terrain types as the same day a year later.

Basically, instead of relying on mapgen to "hard gen" terrain, the terrain changing algorithm will have to replace it on all the tiles where it applies, and do it from the day 1, not after few days.

**Comment by kevingranade** [comment]
Lapsed suggestion thread.

**Comment by BevapDin** [comment]
> . Possibly a ring buffer that stores the last few months of weather

You don't (currently) need that. The weather generator can give you the weather at any time and any place you like. See `weather_generator::get_weather`.

**Comment by kevingranade** [comment]
API wise, this belongs in the weather module, which would grow something like 'int recent_rainfall(start_time, end_time)', it would be up to the weather module to provide it in an efficient way, such as by caching results.
